### PR TITLE
refactor: Migrate to C++20

### PR DESCRIPTION
**Pull request by tibetiroka**
**Refactor**

This PR migrates the source code to C++20.

## Summary
Changes made:
 - Migrated to Catch2 v3, because v2's C++20 support sucks
 - Added Catch2 to vcpkg to avoid having their entire source code in our tree
 - Fixed some compatibility issues

## New features
Though we won't get every single C++20 feature, we can get [most of them](https://en.cppreference.com/w/cpp/compiler_support/20) with the following compilers:
 - GCC 10
 - Clang 13

The oldest supported version of Debian is Buster, going EOL at the end of this month (within a week). The next version already ships GCC 10, so we should be good there. The oldest LTS Ubuntu also ships GCC 10, though the default package only has GCC 9.

macOS 12 ships XCode 13 with Apple Clang 13, which is roughly comparable to Clang 13. (Also, it's going EOL in the next few months as well. The next version comes with XCode 14 at minimum.)

MSVC also has pretty good support, and we aren't tied to version numbers there for support as far as I know.

## Usage examples
Some nice C++20 features we will immediately be getting (in the order listed in the [official docs](https://en.cppreference.com/w/cpp/compiler_support/20)):
 - Lambda captures with `[=, this]`
   - In fact, the old implicit `this` capture is now deprecated.
 - Designated initializers
   - It's like a constructor but not quite... I'm not sure it's going to be too useful for us, since we generally mandate constructors anyway.
 - Concepts
   - Really nice for complex templates, not so useful anywhere else
 - **Three-way comparison operator**
   - This would allow us to nicely extract flat map functionality from Dictionary, since it uses three-way comparison (strcmp) internally.
 - Even nicer structured bindings
 - **`std::map::contains`** (also for other containers)
 - **`std::erase`, `std::erase_if`**
   - Took them long enough. Comes with execution policy support, too.
 - **`unseq` execution policy**
   - Possibly allows for vectorization without random-access iterators (loops over `std::list`, for instance)
 - Various mathmatical constants
   - Pi.h is now less useful (we still need the deg/rad conversion constants, though)

## Testing Done
It compiles.

### CODE REVIEW

**Comment by tibetiroka** [comment]
Waiting on #10218.

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -8,10 +8,27 @@ if(NOT MSVC)
...
+elseif(UNIX AND APPLE)
+	set(CUSTOM_PATH "vcpkg_installed/${VCPKG_TARGET_TRIPLET}")
+else()
+	if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+		set(CUSTOM_PATH "build/ci/vcpkg_installed/x64-windows-release")
+	else()
+		set(CUSTOM_PATH "build/ci/vcpkg_installed/win64-release")
+	endif()
+endif()
+set(CMAKE_PREFIX_PATH "${CMAKE_SOURCE_DIR}/${CUSTOM_PATH}")
```
Comment (petervdmeer) [inline tests/CMakeLists.txt:25]: Probably just my lack of understanding, but why do we need to specify a tool-chain file and custom paths here?

Comment (tibetiroka) [reply tests/CMakeLists.txt:25]: I'm honestly not sure. It's supposed to "just work", but it didn't, and this was the solution recommended to me... I'm not clear on why it didn't import the packages correctly by default.

Comment (petervdmeer) [reply tests/CMakeLists.txt:25]: I feel we need to understand this a little bit better than we currently do. CMake has very smart caching systems and if this somehow affects the caching systems, then we will lose quite some efficiency benefits from CMake.

Comment (tibetiroka) [reply tests/CMakeLists.txt]: I think it's fixed now.

**Comment by TomGoodIdea** [comment]
Seems to work on Windows Vista, nice

**Review by petervdmeer — APPROVED** [review]

**Comment by petervdmeer** [comment]
Reviewed; didn't see anything strange in the changed files.

**Pull request merged**